proof-of-work/lightweight-check/src/HSChain/POW.hs

```diff
@@ -0,0 +1,138 @@
+      order' -> cmp order' by1s by2s -- non-equality at higher byte overrides ordering.
+    cmp  _      _          _         = error "bytestring must be of same length."
+
+-- |Decode clause from the hash.
+--
+-- We are using here the same encoding as in the C code, which is
+-- variable index must be in 1..256 (instead of 0..255) and literal
+-- polarity is encoded in the sign.
+--
+-- This way it is slightly easier to validate against C code.
+--
+-- Also
+createClause :: ByteString -> [Int]
+createClause currentHash = clause
+  where
+    clause = take clauseLiteralsCount_K $ decodeClause [] $ B.unpack currentHash
+    decodeClause acc [] = []
+    decodeClause acc (bLo:bHi:bs) = literal : decodeClause (variable : acc) bs
+      where
+        w16 = fromIntegral bLo + fromIntegral bHi * 256 :: Word16
+        polarity = w16 .&. 1
+        assignTrue = polarity == 1
+        modulo = answerSize * 8
+        randomIndex = ((fromIntegral $ shiftR w16 1) `mod` modulo) + 1 -- 1..256
+        variable = skipUsed acc randomIndex
+        skipUsed ps index
+          | elem index ps = skipUsed ps nextIndex
+          | otherwise = index
+          where
+            nextIndex = mod index modulo + 1 -- index is in range 1..256; taking modulo wraps its value to 0..256 (but next one); adding 1 move it back into 1..256 range.
+        literal = if assignTrue then variable else negate variable
+
+-- |Check whether answer solves puzzle for header.
+solvesPuzzle :: Int -> ByteString -> ByteString -> Bool
+solvesPuzzle numClauses answer header = all clauseSatisfied clauses
+  where
+    hashByStrToByStr byStr = toByStr
+      where
+        Hash toByStr = hashBlob byStr :: Hash SHA256
+    clauses = take numClauses $ map createClause $ tail $ tail $ iterate hashByStrToByStr header
+    clauseSatisfied literals = satisfied
+      where
+        satisfied = any literalSatisfied literals
+        literalSatisfied l = (l > 0) == answerBitTrue
+          where
+            v = abs l
+            bit = v - 1
+            (byte, bitInByte) = divMod bit 8
+            answerByte = B.index answer byte
+            answerBitTrue = (answerByte .&. shiftL 1 bitInByte) /= 0
+
+check :: ByteString -> ByteString -> ByteString -> POWConfig -> IO Bool
+check headerWithoutAnswer answer hashOfAnswerHeader POWConfig{..}
+  | B.length hashOfAnswerHeader /= answerSize = return False -- wrong data size.
+  | targetCompareResult == LT = return False -- final hash not under target complexity.
+  | computedHash /= hashOfAnswerHeader = return False -- hashes do not match.
+  | otherwise = return (solvesPuzzle powCfgClausesCount answer headerWithoutAnswer)
+  where
+    encodedTarget = encodeIntegerLSB powCfgTarget
+    targetCompareResult = compareLSB encodedTarget hashOfAnswerHeader
+    Hash computedHash = hashBlob $ B.concat [answer, headerWithoutAnswer] :: Hash SHA256
+    
+--  B.useAsCStringLen headerWithoutAnswer $ \(hdr, hdrLen) -> 
```

Comment:
It's probably better to just remove C-based code and just reference corresponding function in full-minig package. It will go out of date